### PR TITLE
Update pyproject.toml

### PR DESCRIPTION
**Pull request by VladimirFokow**
tests passed, [can be merged](https://github.com/pypa/sampleproject/pull/204#issuecomment-1971882970)✅

<br />

`pyproject.toml` is outdated. I have:

- moved the `[build-system]` to the start of the file (because this setting is pretty important, and I saw this everywhere in examples in the docs)

- simplified & updated the `[build-system]` table

- removed all `# Optional` comments (because the majority of keys are optional by default) - it's better to mark only the `# REQUIRED` ones

- added useful explanations and links

(plus, integrated the changes from pull request #197)

<br />

## Explanations:

> If you are happy with the current edits, **no need to read all this!**

> If you're not, here are my motivations for EACH change.
> Note:
> These changes improve the understandability for the end user (like me) a lot. And they fix all outdated things. This PR has one goal: to update `pyproject.toml`.

<br />

<details>
  <summary>click: Explanations</summary>

- How I changed the `[build-system]` table:

  The `[build-system]` parameters that I used here are: 
    > ```toml
    > [build-system]
    > requires = ["setuptools"]
    > build-backend = "setuptools.build_meta"
    > ```
  - why `setuptools`? - it is the default build backend
  - why these parameters? - this is the minimal valid example of parameters that I could find [in the Docs](https://setuptools.pypa.io/en/latest/build_meta.html#how-to-use-it)
  
  Pros of these new `[build-system]` parameters:
  - Move away from specifying`the assumed default build requirements from pip`:
    - the [current](https://pip.pypa.io/en/stable/reference/build-system/pyproject-toml/#fallback-behaviour) "assumed default build requirements from pip" differ from the ones specified in this file. Fact: this repo is rarely maintained. Minimal example -> no need to maintain so often -> no inaccuracy.
    - the default pip parameters - are needed for legacy compatibility. Instead of recommending the (legacy) default pip build parameters, this file can have the minimal default *recommendation* for new projects. 
  
  Cons of these new  `[build-system]` parameters:
  - doesn't specify the min. version of "setuptools" -> an error will NOT (?) be reported if the available version of `setuptools` is not recent enough. -- (Is this a big problem?) Maybe some min. version should be specified - which one?

<br />

- Why I removed this link: https://pip.pypa.io/en/stable/reference/pip/#pep-517-and-518-support 
  - It's outdated ("This page has moved", now it redirects to a page with no useful info for this purpose), 
  - plus, PEPS 517 and 518 are concerned with supporting a [legacy method](https://pip.pypa.io/en/stable/reference/build-system/setup-py/#setup-py-legacy),
so mentioning them in this sample project (which is a recommendation for NEW projects) is not necessary.

<br />

- Why I removed the "# Optional" comments:
The only required keys are: `name` and `version`. All the other ones are optional.
So it's better to not mark every key as Optional but instead only to mark these 2 as `# REQUIRED`.
    - no future contributors forget to mark something as optional
    - the user understands the general rule (that the majority of keys are optional)
    - the user will check the up-to-date list in a linked specification - so no need to maintain these tags (which is a proved problem: it went for 6 months just now with a PR hanging for fixing the incorrect labels)
    - DRY principle

</details>

---

<details>
  <summary><h3>Other quotes and proofs</h3></summary>

<br />

- [proof](https://packaging.python.org/en/latest/specifications/pyproject-toml/) why there is no `# REQUIRED` comment next to the `[build-system]` table:

> A pyproject.toml file may be used to store configuration details other than build-related data and thus lack a [build-system] table legitimately.

<br />

- [quote](https://packaging.python.org/en/latest/guides/modernize-setup-py-project/#should-pyproject-toml-be-added)
> Should pyproject.toml be added?
> A pyproject.toml file is strongly recommended. The presence of a pyproject.toml 
> file itself does not bring much. What is actually strongly recommended is the 
> [build-system] table in pyproject.toml.

<br />

- [quote](https://packaging.python.org/en/latest/tutorials/packaging-projects/#choosing-a-build-backend)
> Choosing a build backend:
> Both of these values will be provided by the documentation for your build backend, 
> or generated by its command line interface.

<br />

- [quote](https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#declaring-the-build-backend)
> Usually, you’ll just copy what your build backend’s documentation suggests.
> There should be no need for you to customize these settings.

</details>

---

<details>
  <summary><h3><s>Please correct this if it's wrong</s> <a href="https://github.com/pypa/sampleproject/pull/204#issuecomment-1925464397">done </a></h3>✅ </summary>

What comment is suitable if the `build-backend` parameter is missing?

About the `build-backend` parameter:
> If not defined, then the fall-back behaviour is to use setuptools.

I've found this in [only one place](https://setuptools.pypa.io/en/latest/build_meta.html#what-is-it) - in a note in parenthesis; I have doubts: if `pyproject.toml` exists and `requires = NOT setuptools` but `build-backend` isn't defined, will `setuptools` still be used? - or the chosen backend?

On the [other page](https://packaging.python.org/en/latest/specifications/pyproject-toml/#declaring-build-system-dependencies-the-build-system-table), I've seen:

> If the table is specified but is missing required fields then the tool should consider it an error

So the other build backend chooses what to do? - then it's not `setuptools` in this case?

> ..unrelated, but still:
> > Tools should not require the existence of the [build-system] table
> 
> \- I'm not knowledgeable enough to know, how it is possible to use another build backend if the [build-system] doesn't exist.. Does it make sense to write here that `[build-system]` is optional? Can any build systems exist that can work without the `[build-system]` table?

You can just remove this line please:

"# If not defined, then the fall-back behaviour is to use setuptools."

and maybe give this key a "# REQUIRED" comment.. or not?

---

Plus, maybe add this as well?
> Note that the presence of a pyproject.toml file (even if empty) triggers pip to change its default behavior to use build isolation.
https://packaging.python.org/en/latest/guides/modernize-setup-py-project/#where-to-start

<br />

- [quote](https://setuptools.pypa.io/en/latest/build_meta.html#what-is-it) what the default build backend is:

>  if there is no pyproject.toml file or the build-backend parameter is not defined, then the fall-back behaviour is to use setuptools.

</details>

### CODE REVIEW

File: pyproject.toml
Comment (webknjaz) [inline pyproject.toml:6]: It does bring a lot, actually. Many tools take the presence of `[build-system]` as an explicit opt-in for switching to the isolated PEP 517 mode (usually the ones that used to support invoking setuptools directly in the past through calling `setup.py`).

Moreover, if the `build-backend` is unspecified, the `setuptools.build_meta:__legacy__` is the standardized fallback. And it works differently than the modern/explicit `setuptools.build_meta`.

File: pyproject.toml
Comment (webknjaz) [inline pyproject.toml:21]: Most people around here default to the US spellings, not UK. Plus this is not fully accurate.
```suggestion
# If not defined, then the fall-back behavior is to use
# the setuptools' outdated method of operation implemented as
# a `setuptools.build_meta:__legacy__` interface on the
# build backend level.
```

File: pyproject.toml
Comment (webknjaz) [inline pyproject.toml:25]: Although, not explicitly called out, I'm pretty sure it's optional in practice.

File: pyproject.toml
Comment (webknjaz) [inline pyproject.toml:35]: Same here — it's recommended but optional in practice.

**Comment by webknjaz** [comment]
@VladimirFokow almost everything you changed is really optional. Also, I recommend making small atomic changes so that they could be reviewed more easily and accepted faster, without implying countless rounds of reviews.

File: pyproject.toml
Comment (VladimirFokow) [reply pyproject.toml:6]: then also the docs should be changed? because I was taking all info straight from the docs

**Comment by webknjaz** [comment]
@di could you approve the CI run? I don't have that privilege in this repo.

**Comment by VladimirFokow** [comment]
sorry for many changes.. The total file isn't large though; if people now add their commits on top, we can all move closer towards a good final file.

- yes, everything is optional, so maybe it's easier to say this one time, and only mark the required stuff as required?) this is the last point in my "Notes and info for the reviewers:"

File: pyproject.toml
Comment (VladimirFokow) [reply pyproject.toml:21]: thanks!

just a small question:

On the [other page](https://packaging.python.org/en/latest/specifications/pyproject-toml/#declaring-build-system-dependencies-the-build-system-table), I've seen:

> If the table is specified but is missing required fields then the tool should consider it an error

So the other build backend chooses what to do? - then it's not `setuptools` at all in this case?

**Comment by webknjaz** [comment]
> The total file isn't large though; if people now add their commits on top, we can all move closer towards a good final file.

Yes, which may mean months in-between reviews. Each asking to change one line. And the entire patch not being acceptable for a year.

It's not about the size of the patch but semantically different changes. Does this PR mean to apply arbitrary updates that aren't related to each other? That doesn't sound like a sustainable strategy to me. Combining formatting and functional changes in the same patch usually makes it way easier for the reviewers to get lost in what's happening, which in turn prompts them to postpone doing reviews or delegate to somebody brave.

I have a few collections or links related to contributing PRs. Perhaps, you'll find some of them useful: https://gist.github.com/webknjaz/2aa9c7a43b51c1385260ff87e0cbb9e3 / https://gist.github.com/webknjaz/a7362787a80067af8621a85a71746ca1 / https://gist.github.com/webknjaz/cb7d7bf62c3dda4b1342d639d0e78d79 (and https://gist.github.com/webknjaz/a5a4fb374b7579de827e6bedb93a5220 which is mostly dedicated to advanced Git usage, so maybe isn't applicable to you).

> * yes, everything is optional, so maybe it's easier to say this one time, and only mark the required stuff as required?) this is the last point in my "Notes and info for the reviewers:"

I don't know. Other people may have opinions. I only wanted 9 bytes removed from this file.

File: pyproject.toml
Comment (webknjaz) [reply pyproject.toml:6]: Do you have a specific link? Did you copy it as is? Did you change it taking the context into account?

Comment (webknjaz) [reply pyproject.toml:6]: Ah. Found it in the PR description. It would be useful to add such links right in the diff, next to each change in the patch to eliminate the mental overhead of matching the PR description with the patch that is displayed on a different PR tab.

Comment (webknjaz) [reply pyproject.toml:6]: Hey @sinoroc, you wrote that text IIRC. Could you take a look and consider clarifying the PyPUG document?

File: pyproject.toml
Comment (chrysle) [reply pyproject.toml:35]: It's definitely worth it when publishing to PyPI (which should cover most cases).

File: pyproject.toml
Comment (VladimirFokow) [reply pyproject.toml:6]: :)

I'm not against including the links right in the file! 
(I thought I'd find opposition here due to overcrowding it with links)

Imagine how I feel when there are 7 different places to read about python packaging)

File: pyproject.toml
Comment (webknjaz) [reply pyproject.toml:21]: The `[build-system]` table is consumed by build *frontends*. `setuptools` is a *backend* and wouldn't normally access it. The frontend creates an isolated virtualenv for building the project and pre-provisions it with build dependencies. For that, it consults the `requires` field in order to pre-install some packages into the temporary build environment (there are actually 3 other PEP 517 / PEP 660 hooks that can be consulted for *additional build deps*). Only when the ephemeral build environment is ready, the frontends go ahead and call the hooks defined in the backend that is specified or defaulted to.

File: pyproject.toml
Comment (webknjaz) [reply pyproject.toml:35]: It is recommended, but PyPI does not reject uploads if it's missing.

I suppose the only practical way of knowing if something would be rejected is to run `twine check --strict` against a dist. That's your source of truth for what's really optional in the metadata. Note that backends may generate some metadata fields without consulting the configuration file (as a fallback, for instance).

File: pyproject.toml
Comment (webknjaz) [reply pyproject.toml:6]: It's not about including the links in the code comments. It's about annotating your pull request with context. Using inline discussion threads, like this one.

**Comment by jeanas** [comment]
> I've found this in [only one place](https://setuptools.pypa.io/en/latest/build_meta.html#what-is-it) - in a note in parenthesis; I have doubts: if pyproject.toml exists and requires = NOT setuptools but build-backend isn't defined, will setuptools still be used? - or the chosen backend?

It is setuptools. This is legacy behavior that originates in the history of these fields — `build-system.requires` was initially defined (in PEP 518) before `build-system.build-backend` (which is from PEP 517). Therefore you may encounter rare (?) `pyproject.toml` files with only `build-system.requires`. This information is on https://packaging.python.org/en/latest/specifications/pyproject-toml/, but I don't think we should put it anywhere else.  Users want to know what they should do, not the finer details of how tools keep compatibility with deprecated practices.

> On the [other page](https://packaging.python.org/en/latest/specifications/pyproject-toml/#declaring-build-system-dependencies-the-build-system-table), I've seen:
>
> > If the table is specified but is missing required fields then the tool should consider it an error
>
> So the other build backend chooses what to do? - then it's not setuptools in this case?

The build *frontend* should raise an error. No build backend is invoked.

------------------

General feedback:

> Imagine how I feel when there are 7 different places to read about python packaging)

For this very reason, I would honestly just remove the descriptions of the fields and only put a link to a subheading on the `pyproject.toml` guide above each field (e.g., https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#python-requires).

**Comment by webknjaz** [comment]
> **Edit:** However, if build isolation is used (the default in pip) and `setuptools` is not in the `requires` list then this will simply give an error about not being able to import the backend.

I remembered I was updating a related explanation in pip docs recently (https://github.com/pypa/pip/pull/12449) and here's exactly what it does: https://pip.pypa.io/en/stable/reference/build-system/pyproject-toml/#fallback-behaviour. The exact implementation of the decision tree is @ https://github.com/pypa/pip/blob/811beab/src/pip/_internal/pyproject.py#L29-L179.

**Comment by webknjaz** [comment]
> I would honestly just remove the descriptions of the fields

This has its ins and outs. My understanding is that people may use this repo as a starting point and would be filling out the files as templates. In such a case, keeping the context close and not sending them on the round trip to the external docs has the least mental overhead. OTOH, keeping the descriptions in sync and up-to-date causes a burden for the maintainers.

I'm not going to make a judgement call, but as this repo doesn't get a lot of activity, it does feel like reducing said maintenance burden would be beneficial in the long run.

**Comment by VladimirFokow** [comment]
I actually didn't mean that the descriptions in this file were a burden - I think the descriptions here are good.

That said, I also don't mind removing them completely if we have a specification link with them all.

Thinking about the larger picture -

I think in this file specifically, there can be quick comments about why we decided to include these specific keys for the example project, even though they are optional, they are good to have, because `...`

**Comment by webknjaz** [comment]
> I actually didn't mean that the descriptions in this file were a burden - I think the descriptions here are good.

You're talking from the PoV of a consumer / end-user. We're taking a point of a maintainer who ends up having to update those description regularly. It's a kind of a responsibility that tends to pile up and contribute to people burning out.

**Comment by VladimirFokow** [comment]
To sum up:

I can create a commit to:

- fix spelling: "behavior"

- For the `build-backend`, can use a simple comment: 
`# If not defined, then legacy behavior can happen.`

---

Create a new issue to:
  - choose the set of keys as a default quick start
  - have only those descriptions here that are already in the docs (or summarise the docs)

(so if we want to add something new here - add it to the docs first, then to here)

  - remove comments if: obvious / unnecessary / too long
  - add useful quick summarisations to this file from the docs

---

Does everyone agree for me to do these 2 actions?

File: pyproject.toml
Comment (VladimirFokow) [reply pyproject.toml:6]: So from the user point of view:

`[build-system]` table exists --(implies)--> `build-backend` is specified 
(because of the comment `# REQUIRED if [build-system] table is used`)

Then the user might think, "and what if I don't include the `[build-system]` table?"
And the importance of having a `[build-system]` table is correctly stated in the comment:
```py
# The presence of a pyproject.toml 
# file itself does not bring much. What is actually strongly recommended is the 
# [build-system] table in pyproject.toml
```

---

Do you agree that "the presence of a pyproject.toml file itself (i.e. without the `[build-system]` table) does not bring much"? We can remove only this sentence if it's incorrect

**Comment by pfmoore** [comment]
> Does everyone agree for me to do these 2 actions?

I don't think we need a PR to change the spelling of "behaviour". There's no policy that American spelling should be used over British spelling, so leave existing spellings alone (although I thought the spelling correction was just a change to something you'd added in this PR, not a change to existing text?)

As for the second action, I have the same reservations as @webknjaz - I don't think it will add enough benefit to justify the maintenance effort. Simply reading the discussion on this PR has been more work on this repository than I've otherwise done in months - this project is supposed to be (very) low maintenance. But I'm largely indifferent to the idea, if you want to make a PR and someone wants to merge it, I won't object. I just don't *personally* think it's that important.

**Comment by VladimirFokow** [comment]
@pfmoore "behaviour" is original, "behavior" was suggested.

Anyway, I just want to improve the current `pyproject.toml` as it's in an unacceptable condition for me. 
I am happy to quickly apply a minimum set of changes that everyone agrees on.

**Comment by sinoroc** [comment]
Thanks a lot @VladimirFokow for working on this pull request :)

File: pyproject.toml
Comment (sinoroc) [reply pyproject.toml:6]: I have been mentioned, so here is my take:

I did not understand what is the "issue" with the sentence. Seems fine to me. `pyproject.toml` in itself does not do much (or at least it should not do much, at some point in time its presence changed the behavior of things for other tools but I think it has been fixed for a wile now).

To be honest I do not think we need all these prose in there. I guess I would only put the links to the [guide](https://packaging.python.org/en/latest/guides/writing-pyproject-toml/) (now it says "specification" but it is a link to a guide) and then the [specification](https://packaging.python.org/en/latest/specifications/pyproject-toml/).

**Review by sinoroc — APPROVED** [review]

**Review by sinoroc — APPROVED** [review]

**Review by pfmoore — APPROVED** [review]

**Comment by VladimirFokow** [comment]
same check issue that was in #197..

Any ideas what the next steps should be?

File: pyproject.toml
Comment (chrysle) [inline pyproject.toml:132]: How about this? I'm missing an explanation here, the comment seems at this time somewhat content-less.

```suggestion
# Optional dependencies the project provides. These are commonly 
# referred to as "extras". For a more extensive definition see:
# https://packaging.python.org/en/latest/specifications/dependency-specifiers/#extras
```

Comment (VladimirFokow) [reply pyproject.toml:132]: I don't mind adding this.
Waiting for approval from others before making a commit.

For me personally, this link was more helpful to understand what extras are:
https://pip.pypa.io/en/stable/cli/pip_install/#examples
(point 7.)
plus, it also leads to the specification (that you have provided)

**Comment by chrysle** [comment]
> Any ideas what the next steps should be?

The deprecated entry point interfaces for `importlib_metadata` have been removed in v5. I think we should pin that on `< 5` with an environment marker for Python 3.7.

**Comment by VladimirFokow** [comment]
Is [this](https://github.com/python/importlib_metadata/issues/406#issuecomment-1264662847) the issue?:
> I've attempted to engage with the maintainers of that project, but they refuse to engage and dismiss my proposals without suggesting alternatives.

Unfortunately, I don't have the technical skills now to figure out another solution.

---

Python 3.7 has reached its [end-of-life](https://devguide.python.org/versions/) What if we just drop support for Python 3.7?

**Comment by chrysle** [comment]
> Python 3.7 has reached its [end-of-life](https://devguide.python.org/versions/) What if we just drop support for Python 3.7?

Yes, but in a separate PR I'd suggest. For now, it's mostly about pinning the `importlib_metadata` requirement in `tox.ini`:

https://github.com/pypa/sampleproject/blob/3fb1461c7c13b6832ed0f25625b19f2d321a74b7/tox.ini#L28

**Comment by sinoroc** [comment]
> Should some min. version of "setuptools" be specified? If so, which one?

I would say no.

It might trigger a whole (long) discussion of its own, so I would not push too hard on this topic within this pull request if I were you.

**Comment by jeanas** [comment]
> > Should some min. version of "setuptools" be specified? If so, which one?
>
> I would say no.
>
> It might trigger a whole (long) discussion of its own, so I would not push too hard on this topic within this pull request if I were you.

IMHO, this is easy to answer: the setuptools documentation [does not have a bound](https://setuptools.pypa.io/en/latest/userguide/quickstart.html#basic-use), so this sample shouldn't either.

**Comment by VladimirFokow** [comment]
can be merged✅

**Review by di — APPROVED** [review]

**Comment by webknjaz** [comment]
> > Should some min. version of "setuptools" be specified? If so, which one?
> 
> I would say no.

The same considerations should be applied as with normal software deps. If a project uses a specific feature from a lib, it's natural to declare that a minimum version of that lib is required. Similarly, if there are features used in the packaging config that setuptools only supports starting with a certain version, it's good to communicate that. I recommend pairing that with an explanatory comment like `# feature X was first implemented in setuptools vY` and so on.

But outside of such context, when the features used are implemented by any version of a dependency, then the dependency declaration should communicate exactly that — an unbound requirement specifier.

Upper bounds are more complicated and shouldn't generally be used most of the time.

**Pull request merged**
